File: LeetCode/p45.cpp
class Solution {
public:
    int jump(const vector<int> &nums) {
        auto sz = nums.size();
        if (sz <= 1)
            return 0;
        int this_cover = 0;
        int times = 0;
        for (int i = 0;; ++times) {
            int next_cover = this_cover;
            if (next_cover >= sz - 1)
                return times;
            for (int j = i; j <= this_cover; ++j) {
                next_cover = max(j + nums[j], next_cover);
            }
            i = this_cover + 1;
            this_cover = next_cover;
        }
        return -1;  // never
    }
};

File: LeetCode/p45.rs
impl Solution {
  pub fn jump(nums: Vec<i32>) -> i32 {
    let mut ret = 0;
    let mut this = 0;
    let mut next = 0;
    for i in 0..nums.len() {
      if i > this {
        ret += 1;
        this = next;
        if next >= nums.len() - 1 {
          break;
        }
      }
      next = next.max(i + nums[i] as usize);
    }
    ret
  }
}

